### ISSUE TITLE
Native file chooser messes with main window resizing

### ISSUE THREAD

**Issue by MoAlyousef** [fltk/fltk#401]
Hello
Native file chooser removes maximize button from main window's decoration on linux (tested on a kubuntu machine), and disables resizing on xfce. 

Given this minimal repro:
```cpp
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Native_File_Chooser.H>

void but_cb(Fl_Widget *, void *) {
    auto dlg = new Fl_Native_File_Chooser();
    dlg->show();
}

int main() {
    auto win = new Fl_Window(400, 300, "Window");
    auto button = new Fl_Button(160, 200, 80, 40, "Click");
    win->end();
    win->resizable(win);
    win->show();
    button->callback(but_cb, NULL);
    return Fl::run();
}
```
The main window appears with a maximize button, after clicking the button and spawning the dialog, the decoration appears to modify where the maximize button disappears and the close button shifts towards the left a bit. Taking a screenshot is proving to be a challenge on my Kubuntu machine, but this was reported in fltk-rs:
https://github.com/fltk-rs/fltk-rs/issues/1146

(on xfce on arch linux, resizing is also affected)

The change first appeared after this commit:
https://github.com/fltk/fltk/commit/0f335ab4409943ff59dd0c542e9a3c9947c911d7

Edit:
Ok got some screenshots using Spectacle:
![image](https://user-images.githubusercontent.com/37966791/155872795-8fd53723-4120-40c6-9aaf-4ce3e3ce06ec.png)

After:
![image](https://user-images.githubusercontent.com/37966791/155872806-eb032337-e263-4650-9564-e39e16bcbdc5.png)

**Comment by Albrecht-S** [comment]
Thanks for reporting this. I don't have a KDE system available but I believe that @ManoloFLTK can test and hopefully fix this. Manolo?

**Comment by MoAlyousef** [comment]
@[ThiNei2l](https://github.com/ThiNei2l), the original reporter, also added:
> I can reproduce your behaviour: on plasma only the button disappears, but resizing using the window border still works.
> 
> What I have seen for now: LXQt / openbox and MATE / marco: Everything works fine. Plasma / KWin: Only the button disappearing. Xfce / Xfwm: Resizing completely broken.

**Comment by Albrecht-S** [comment]
@MoAlyousef FLTK's native filechooser tries to detect which filechooser is available and uses either Gtk or KDE or its own filechooser. Please ask the OP which filechooser was raised in their tests. According to the description so far and the mentioned commit 0f335ab4409943ff59dd0c542e9a3c9947c911d7 I assume that the faulty behavior is caused only by the KDE filechooser. Can you or the OP confirm this?

**Comment by MoAlyousef** [comment]
I've asked them the question. But you're right, it could be a later commit since xfce doesn't run kdialog by default. But the change in dialog behavior appeared after 14/1/2022 (since fltk-rs version 1.2.27 (which doesn't exhibit the behavior) was released in 14/1/2022).

**Comment by Albrecht-S** [comment]
I think the mentioned commit 0f335ab440 is indeed the culprit. It's the first one that messes with unsetting and resetting the size_range() of windows while the filechooser dialog is running. I can confirm similar behavior with the Gtk filechooser - the code has been copied from the Gtk filechooser wrapper to the Kdialog filechooser wrapper in that commit.

Note: I still can't test the KDE filechooser but I could see an effect (improvement) in the Gtk filechooser after inserting `Fl::flush()` after resetting the `size_range()`. Can you please test if the following patch changes anything?
```diff
diff --git a/src/Fl_Native_File_Chooser_Kdialog.cxx b/src/Fl_Native_File_Chooser_Kdialog.cxx
index 368a2b6ed..b374623fa 100644
--- a/src/Fl_Native_File_Chooser_Kdialog.cxx
+++ b/src/Fl_Native_File_Chooser_Kdialog.cxx
@@ -165,6 +165,7 @@ int Fl_Kdialog_Native_File_Chooser_Driver::show() {
       first_dim = dim->next;
       delete dim;
     }
+    Fl::flush();
     if (data.all_files) {
       // process text received from pipe
       if (data.all_files[strlen(data.all_files)-1] == '\n') data.all_files[strlen(data.all_files)-1] = 0;
```

**Comment by Albrecht-S** [comment]
I could now try it with KDE as well and it looks like this is a "KDE feature" - or whatever window manager is involved, i.e. it's a window manager issue. Maybe FLTK can do something about it but I don't know what could be done. I wrote a small test program that shows the same effect without running a filechooser:
```C++
// Fl_Window::size_range() test program for the Fast Light Tool Kit (FLTK).
//
// Author:    Albrecht-S
// Date:      Feb 28, 2022
// Copyright: Public Domain, use as you like
// Build:     fltk-config --compile size_range.cxx

#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>

const double timer = 10.0;
int resizable = 1;

void resizable_cb(void *v) {
  Fl_Window *win = (Fl_Window *)v;
  resizable = 1 - resizable;
  if (resizable)
    win->size_range(100, 100, 0, 0);
  else
    win->size_range(200, 200, 200, 200);
  printf("Window (%3d, %3d, %3d, %3d): resizable %s\n",
         win->x(), win->y(), win->w(), win->h(),
         resizable ? "ON" : "OFF");
  fflush(stdout);
  Fl::repeat_timeout(timer, resizable_cb, (void *)win);
}

int main(int argc, char **argv) {
  Fl_Window *window = new Fl_Window(340, 180, "win1");
  Fl_Box *box = new Fl_Box(20, 20, 300, 140, "Hello!");
  box->box(FL_UP_BOX);
  box->labelsize(24);
  window->end();
  window->resizable(box);
  Fl::add_timeout(timer, resizable_cb, window);
  window->show(argc, argv);
  return Fl::run();
}
```

What behavior should be expected? This program toggles the size_range() settings every 10 seconds. On my Linux Mint Cinnamon this works as expected, on Debian with Gnome this works fine as well, but when I run it under KDE it doesn't show the minimize/maximize/close buttons as it should do.

For now this is only a test program, I'd appreciate reports about its behavior on different systems and WM's.

**Comment by ThiNei2l** [comment]
Hello,

I stumbled upon this issue using fltk-rs.
Just tried the above program. After disabling and then re-enabling resizing I can observe the same behaviour that I see after opening a file chooser:

Under LXQt and MATE everything is fine.
With KDE the maximize button vanishes but I can still resize the window using it's borders.
On Xfce, not only the button disappears but resizing is completely impossible.

**Comment by Albrecht-S** [comment]
@ThiNei2l Thanks for confirmation. Dealing with different WM's is always difficult and I don't know if it's the WM's fault or if we can improve something. We'll look further into it.

**Comment by ManoloFLTK** [comment]
I see that the problem can be fixed, with KDE, by hiding/showing the window just before
the size_range() call intended to restore its buttons to the window titlebar. In other words
add
`win->hide(); win->show()`
just before line 164 of file src/Fl_Native_File_Chooser_Kdialog.cxx fixes the problem under KDE.

However, since other window managers may fail, I would tend to prefer to remove altogether
the code that removes and then restores the titlebar fullscreen button. The objective of that code is to prevent  a user from doing:
1) open a file chooser
2) click on the previous FLTK window which passes in front of the chooser window
3) maximize the FLTK window --> the FLTK window is inactive and the active window, the chooser, is behind it, invisible.

Without this code, the FLTK window is inactive, except it can be minimized which is not bad, and maximized, which creates the problem described above. It can't be closed, which is very good.

@Albrecht: what do you think?

**Comment by ManoloFLTK** [comment]
@Albrecht-S : My proposal: 
- make code that hides and then restores the window maximize button conditional on a boolean returned by a virtual member function of class `Fl_Unix_System_Driver`;
- have this member function return `false` under X11 and `true` under Wayland.

**Comment by Albrecht-S** [comment]
@ManoloFLTK: Here are my proposals:

1. Remove the code that fiddles with the maximize button entirely. This could fail and even crash if **any** top level window is deleted while the filechooser is active. This is possible for instance via a timeout (I can provide a working example that crashes).
2. Do this for both the KDE and the Gtk filechoosers.
3. Do not make this code conditional as suggested in your second comment https://github.com/fltk/fltk/issues/401#issuecomment-1055319659.

I'm also wondering why we need to add the special dispatch function to disable window interaction. Wouldn't it be easier and more effective to just `deactivate()` the window?

**Comment by Albrecht-S** [comment]
**_Note:_** Altough the proposals above would fix the filechooser issues this wouldn't fix the interaction with the WM as shown in my demo program above. As said before I don't know if we can fix this at all or if this needs to be fixed in the WM. However, this is another issue that should be investigated separately.

**Comment by Albrecht-S** [comment]
> I'm also wondering why we need to add the special dispatch function to disable window interaction. Wouldn't it be easier and more effective to just deactivate() the window?

I just tested this. It's different than opening a modal window (which doesn't gray out the widgets) but it works as intended and it avoids subtle problems as seen with trying to deactivate the resize/maximize buttons.

**Comment by Albrecht-S** [comment]
@ManoloFLTK You wrote:

> Without this code, the FLTK window is inactive, except it can be minimized which is not bad, and maximized, which creates the problem described above. It can't be closed, which is very good.

My suggestion above would allow the user to close the window but this can easily avoided (if wanted) by overwriting the window callback (again, I tested this and I have a working proof of concept).

**Comment by Albrecht-S** [comment]
Hmm, I found out that `Fl_Window` widgets can't be deactivated (for whatever reason, it's documented behavior). Hence my suggestion above wouldn't work this way. I'm open for other proposals...

**Comment by ManoloFLTK** [comment]
OK. Since window->deactivate() is not solution, I propose to keep the condition code while protecting memorized Fl_Window pointers by Fl_Widget_Tracker objects to protect from possible window deletion during the event loop. That's now committed in the Wayland branch.

**Comment by ManoloFLTK** [comment]
Should be fixed at d8cb8ec .
